File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int dp[1001][1001];
    int helper(string &s, string &t, int m, int n){
        
        if(m == 0 || n == 0)    return 0;
        
        if(dp[m][n] != -1)   return dp[m][n];
        
        if(s[m - 1] == t[n - 1])
            return dp[m][n] = 1 + helper(s, t, m - 1, n - 1);
        
        int ans1 = helper(s, t, m - 1, n);
        int ans2 = helper(s, t, m, n - 1);
        
        return dp[m][n] = max(ans1, ans2);        
    }
    
    int longestCommonSubsequence(string text1, string text2) {
        
        int m = text1.size();
        int n = text2.size();
        
        memset(dp, -1, sizeof(dp));
        
        return helper(text1, text2, m, n);
        
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    private int t[][];
    int helper(String s1, String s2, int m, int n){
        if(m == 0 || n == 0)    return 0;
        
        if(t[m][n] != -1)   return t[m][n];
        
        if(s1.charAt(m - 1) == s2.charAt(n - 1))
            return t[m][n] = 1 + helper(s1, s2, m - 1, n - 1);
        
        int a = helper(s1, s2, m - 1, n);
        int b = helper(s1, s2, m, n - 1);
        
        return t[m][n] = Math.max(a, b);
    }
    public int longestCommonSubsequence(String text1, String text2) {
        t = new int[1001][1001];
        
        for(int i = 0; i < 1001; i++){
            Arrays.fill(t[i], -1);
        }
        int m = text1.length(), n = text2.length();
        return helper(text1, text2, m, n);
    }
}